src/routes/api/v1/avatar.rs

```diff
@@ -79,24 +81,57 @@ pub async fn avatar(
 	Ok(fallback_response(
 		params.fallback,
 		"Record not found".to_string(),
+		&config,
 	))
 }
 
-fn fallback_response(fallback: Option<Url>, error_msg: String) -> Response {
+fn fallback_response(fallback: Option<Url>, error_msg: String, config: &Config) -> Response {
 	fallback.map_or_else(
 		|| ErrorResponse::not_found(error_msg).into_response(),
-		|fallback| Redirect::temporary(fallback.as_str()).into_response(),
+		|fallback| {
+			// If whitelist is not set, block all fallback URLs
+			if config.whitelisted_avatar_domains.is_none() {
+				return ErrorResponse::forbidden(
+					"Fallback URLs are not allowed when whitelist is not configured".to_string(),
+				)
+				.into_response();
+			}
+
+			// Check if the fallback URL's domain is in the whitelist
+			if let Some(domain) = fallback.host_str() {
+				let domain = domain.to_lowercase();
+				let whitelist = config.whitelisted_avatar_domains.as_ref().unwrap();
+
+				// Check if the domain exactly matches a whitelisted domain
+				// or is a subdomain of a whitelisted domain
+				if whitelist
+					.iter()
+					.any(|allowed| domain == *allowed || domain.ends_with(&format!(".{}", allowed)))
```

Comment:
```suggestion
					.any(|allowed| domain == allowed || domain.ends_with(&format!(".{}", allowed)))
```
I think rust will automatically deference this for you and it's more idiomatic